src/lib.rs

```diff
@@ -366,14 +450,25 @@ impl<T> Receiver<T> {
     /// Panics if called after this receiver has been polled asynchronously.
     #[cfg(feature = "std")]
     pub fn recv(self) -> Result<T, RecvError> {
+        // Note that we don't need to worry about changing the state to disconnected or setting the
+        // state to an invalid value at any point in this functon because we take ownership of
+        // self, and this function does not exit until the message has been received or both side
+        // of the channel are inactive and cleaned up.
+
         let channel_ptr = self.channel_ptr;
 
         // Don't run our Drop implementation if we are receiving consuming ourselves.
         mem::forget(self);
 
+        // SAFETY: the existence of the `self` parameter serves as a certificate that the receiver
+        // is still alive, meaning that even if the sender was dropped then it would have observed
+        // the fact that we're still alive and left the responsibility of deallocating the
+        // channel to us, so channel_ptr is valid
         let channel = unsafe { channel_ptr.as_ref() };
 
-        match channel.state.load(SeqCst) {
+        // ORDERING: we use acquire ordering to synchronize with the write of the message in the
+        // case that it's available
+        match channel.state.load(Acquire) {
```

Comment:
What's the motivation for putting the `Acquire` ordering here rather than an explicit `fence(Acquire);` down in the `MESSAGE` branch? Only one branch need this ordering, so why pay for it in all branches?

Reply:
There's generally no hard-and-fast rule about when to use a fence vs not. I would classify any potential performance win there as an over-optimization however, because if we end up _not_ needing the acquire barrier, then we will _probably_ end up parking, which will take immensely more wall-clock time. I've also heard that sometimes `load(Acquire)` can be more efficient on some platforms, but I've not been given any specific examples or benchmarks.

Reply:
In briefly talking with some other devs, someone claims to have benched that `load(Acquire)` is faster than `load(Relaxed); fence(Acquire);` on aarch64. I think there might be an argument for getting rid of the fences on the sending side given that those branches are likely to be taken. I'd mostly consider using the fence optimization for branches which are unlikely to be taken.

Reply:
Ok. Yeah, good argument. I was not aware that specifying the ordering in the load could be optimized better. But I guess that makes sense. Would be cool to run those benchmarks or write my own at test it. Do you have any links to anything on this?